File: bsd_3_clause_gen.py
import tkinter as tk


class AppWindow(tk.Tk):
    def __init__(self) -> None:
        super().__init__()
        self.title("BSD 3-Clause License Generator")
        self.geometry("800x600")

        # Configure grid for responsive layout
        self.grid_columnconfigure(0, weight=1)
        self.grid_rowconfigure(1, weight=1)

        # Create input section
        input_frame = tk.Frame(self, bg="#f5f5f5", bd=0)
        input_frame.grid(row=0, column=0, padx=20, pady=20, sticky="ew")
        input_frame.grid_columnconfigure(0, weight=1)

        # Input labels and fields
        tk.Label(input_frame, text="Copyright (c)").grid(
            row=0, column=0, sticky="e", padx=5, pady=5)

        self.year_var = tk.StringVar()
        year_entry = tk.Entry(
            input_frame, textvariable=self.year_var, width=10, font=("Arial", 14))
        year_entry.grid(row=0, column=1, sticky="w", padx=5, pady=5)

        self.author_var = tk.StringVar()
        author_entry = tk.Entry(
            input_frame, textvariable=self.author_var, width=40, font=("Arial", 14))
        author_entry.grid(row=0, column=2, sticky="ew", padx=5, pady=5)
        input_frame.grid_columnconfigure(2, weight=1)

        # Buttons frame
        button_frame = tk.Frame(input_frame, bg="#f5f5f5")
        button_frame.grid(row=1, column=0, columnspan=3, pady=10)

        self.generate_btn = tk.Button(button_frame, text="Generate License", command=self.generate_license,
                                      bg="#007bff", fg="white", font=("Arial", 12), relief="flat", state="disabled")
        self.generate_btn.pack(side="left", padx=5)

        self.copy_btn = tk.Button(button_frame, text="Copy to Clipboard", command=self.copy_to_clipboard,
                                  bg="#007bff", fg="white", font=("Arial", 12), relief="flat", state="disabled")
        self.copy_btn.pack(side="left", padx=5)

        # Output section
        output_frame = tk.Frame(self)
        output_frame.grid(row=1, column=0, padx=20,
                          pady=(0, 20), sticky="nsew")
        output_frame.grid_columnconfigure(0, weight=1)
        output_frame.grid_rowconfigure(0, weight=1)

        # License output text area
        self.license_text = tk.Text(output_frame, font=("Courier New", 11), wrap="word",
                                    bg="#f8f9fa", borderwidth=1, relief="solid")
        self.license_text.grid(row=0, column=0, sticky="nsew")
        self.license_text.insert(
            "1.0", "Please enter the year and author above to generate the BSD 3-Clause License text.")
        self.license_text.config(state="disabled")

        # Scrollbar
        scrollbar = tk.Scrollbar(output_frame, command=self.license_text.yview)
        scrollbar.grid(row=0, column=1, sticky="ns")
        self.license_text.config(yscrollcommand=scrollbar.set)

        # Message label
        self.message_var = tk.StringVar()
        self.message_label = tk.Label(
            self, textvariable=self.message_var, font=("Arial", 12))
        self.message_label.grid(row=2, column=0, padx=20,
                                pady=(0, 10), sticky="w")

        # Event bindings
        self.year_var.trace_add("write", self.validate_inputs)
        self.author_var.trace_add("write", self.validate_inputs)

        year_entry.bind("<KeyPress>", self.validate_year_keypress)

        # Keyboard shortcut for generate (Ctrl+Enter)
        self.bind("<Control-Return>", lambda _: self.generate_license()
                  if self.generate_btn["state"] != "disabled" else None)

    def validate_inputs(self, name: str, index: str, mode: str) -> None:
        import re
        
        _ = name, index, mode
        year = self.year_var.get().strip()
        author = self.author_var.get().strip()

        # Basic validation
        is_valid_year = re.match(r"^\d{4}$", year) is not None
        is_valid_author = len(author) > 0

        if is_valid_year and is_valid_author:
            self.generate_btn.config(state="normal")
        else:
            self.generate_btn.config(state="disabled")

    def validate_year_keypress(self, event: tk.Event) -> str:
        # Only allow numbers and backspace
        if not (event.char.isdigit() or event.keysym == "BackSpace"):
            return "break"
        else:
            return ""

    def generate_license(self) -> None:
        year = self.year_var.get().strip()
        author = self.author_var.get().strip()

        license_text = f"""BSD 3-Clause License

Copyright (c) {year}, {author}
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ""AS IS""
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."""
        self.license_text.config(state="normal")
        self.license_text.delete("1.0", tk.END)
        self.license_text.insert("1.0", license_text)
        self.license_text.config(state="disabled")

        self.copy_btn.config(state="normal")
        self.show_message("License generated successfully!", "success")

    def copy_to_clipboard(self) -> None:
        license_content = self.license_text.get("1.0", tk.END).strip()
        if license_content:
            self.clipboard_clear()
            self.clipboard_append(license_content)
            self.update()  # Update the clipboard
            self.show_message("License copied to clipboard!", "success")
        else:
            self.show_message("No license to copy", "error")

    def show_message(self, text: str, message_type: str) -> None:
        if message_type == "success":
            self.message_label.config(fg="#28a745")
        else:
            self.message_label.config(fg="#dc3545")

        self.message_var.set(text)

        # Clear message after 3 seconds
        self.after(3000, lambda: self.message_var.set(""))


if __name__ == "__main__":
    AppWindow().mainloop()

File: BSD3ClauseGen.cs
#:property PublishTrimmed=false
#:property TargetFramework=net10.0-windows
#:property UseWindowsForms=true

namespace BSD3ClauseGen;

file static class Program
{
    [STAThread]
    internal static void Main()
    {
        Application.EnableVisualStyles();
        Application.SetCompatibleTextRenderingDefault(false);
        Application.Run(new MainForm());
    }
}

file class MainForm : Form
{
    private const string LICENSE_FORMAT = @"BSD 3-Clause License

Copyright (c) {0}, {1}
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

3. Neither the name of the copyright holder nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ""AS IS""
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.";

    private TextBox yearTextBox = null!;
    private TextBox authorTextBox = null!;
    private Button generateButton = null!;
    private Button copyButton = null!;
    private RichTextBox licenseTextBox = null!;
    private Label messageLabel = null!;
    private System.Windows.Forms.Timer messageTimer = null!;

    public MainForm()
    {
        InitializeComponent();
    }

    private void InitializeComponent()
    {
        SuspendLayout();

        Text = "BSD 3-Clause License Generator";
        Size = new Size(800, 600);
        StartPosition = FormStartPosition.CenterScreen;
        MinimumSize = new Size(600, 400);

        var inputPanel = new Panel
        {
            Dock = DockStyle.Top,
            BackColor = Color.FromArgb(245, 245, 245),
            Padding = new Padding(20),
            Height = 100
        };

        var copyrightLabel = new Label
        {
            Text = "Copyright (c)",
            AutoSize = true,
            Location = new Point(20, 25),
            Font = new Font("Arial", 12)
        };

        yearTextBox = new TextBox
        {
            Width = 100,
            Location = new Point(120, 20),
            Font = new Font("Arial", 14),
            MaxLength = 4
        };
        yearTextBox.KeyPress += YearTextBox_KeyPress;
        yearTextBox.TextChanged += ValidateInputs;

        authorTextBox = new TextBox
        {
            Width = 300,
            Location = new Point(230, 20),
            Font = new Font("Arial", 14)
        };
        authorTextBox.TextChanged += ValidateInputs;

        var buttonPanel = new Panel
        {
            Location = new Point(20, 60),
            Height = 30,
            Width = 510
        };

        generateButton = new Button
        {
            Text = "Generate License",
            BackColor = Color.FromArgb(0, 123, 255),
            ForeColor = Color.White,
            FlatStyle = FlatStyle.Flat,
            Font = new Font("Arial", 12),
            Size = new Size(150, 30),
            Location = new Point(0, 0),
            Enabled = false
        };
        generateButton.Click += GenerateLicense;
        generateButton.FlatAppearance.BorderSize = 0;

        copyButton = new Button
        {
            Text = "Copy to Clipboard",
            BackColor = Color.FromArgb(0, 123, 255),
            ForeColor = Color.White,
            FlatStyle = FlatStyle.Flat,
            Font = new Font("Arial", 12),
            Size = new Size(150, 30),
            Location = new Point(160, 0),
            Enabled = false
        };
        copyButton.Click += CopyToClipboard;
        copyButton.FlatAppearance.BorderSize = 0;

        buttonPanel.Controls.AddRange([generateButton, copyButton]);
        inputPanel.Controls.AddRange([copyrightLabel, yearTextBox, authorTextBox, buttonPanel]);

        var outputPanel = new Panel
        {
            Dock = DockStyle.Fill,
            Padding = new Padding(20, 0, 20, 20)
        };

        licenseTextBox = new RichTextBox
        {
            Dock = DockStyle.Fill,
            Font = new Font("Courier New", 11),
            BackColor = Color.FromArgb(248, 249, 250),
            BorderStyle = BorderStyle.FixedSingle,
            ReadOnly = true,
            Text = "Please enter the year and author above to generate the BSD 3-Clause License text."
        };

        outputPanel.Controls.Add(licenseTextBox);

        messageLabel = new Label
        {
            Dock = DockStyle.Bottom,
            Height = 30,
            Font = new Font("Arial", 12),
            Text = "",
            AutoSize = false,
            TextAlign = ContentAlignment.MiddleLeft,
            Padding = new Padding(20, 5, 0, 0)
        };

        messageTimer = new System.Windows.Forms.Timer { Interval = 3000 };
        messageTimer.Tick += MessageTimer_Tick;

        Controls.AddRange([inputPanel, outputPanel, messageLabel]);
        ResumeLayout(false);
    }

    private void YearTextBox_KeyPress(object? sender, KeyPressEventArgs e)
    {
        if (!char.IsDigit(e.KeyChar) && e.KeyChar != (char)Keys.Back)
            e.Handled = true;
    }

    private void ValidateInputs(object? sender, EventArgs e)
    {
        string year = yearTextBox.Text.Trim();
        string author = authorTextBox.Text.Trim();

        bool isValidYear = year.Length == 4 && int.TryParse(year, out _);
        bool isValidAuthor = author.Length > 0;

        generateButton.Enabled = isValidYear && isValidAuthor;
    }

    private void GenerateLicense(object? sender, EventArgs e)
    {
        string year = yearTextBox.Text.Trim();
        string author = authorTextBox.Text.Trim();

        string licenseText = string.Format(LICENSE_FORMAT, year, author);
        licenseTextBox.Text = licenseText;
        copyButton.Enabled = true;

        ShowMessage("License generated successfully!", "success");
    }

    private void CopyToClipboard(object? sender, EventArgs e)
    {
        string licenseContent = licenseTextBox.Text.Trim();

        if (!string.IsNullOrEmpty(licenseContent) &&
            !licenseContent.Contains("Please enter the year and author"))
        {
            Clipboard.SetText(licenseContent);
            ShowMessage("License copied to clipboard!", "success");
        }
        else
        {
            ShowMessage("No license to copy", "error");
        }
    }

    private void ShowMessage(string text, string messageType)
    {
        messageLabel.ForeColor = 
            messageType == "success" ? Color.FromArgb(40, 167, 69) : Color.FromArgb(220, 53, 69);
        messageLabel.Text = text;
        messageTimer.Start();
    }

    private void MessageTimer_Tick(object? sender, EventArgs e)
    {
        messageLabel.ResetText();
        messageTimer.Stop();
    }

    protected override void OnFormClosing(FormClosingEventArgs e)
    {
        yearTextBox.KeyPress -= YearTextBox_KeyPress;
        yearTextBox.TextChanged -= ValidateInputs;
        authorTextBox.TextChanged -= ValidateInputs;
        generateButton.Click -= GenerateLicense;
        copyButton.Click -= CopyToClipboard;
        messageTimer.Tick -= MessageTimer_Tick;
        base.OnFormClosing(e);
    }
}